### PR TITLE
 Fix #22: add pagination and tag filtering to article list 

### LINKED ISSUE #22: Add pagination and tag filtering to homepage and profile pages.
This ticket is to make pagination and tag filtering work. Tags only exist on the home page.

### PR DESCRIPTION
**Pull request by blopker**
Using Frames to keep the article list reusable and the parent pages fast!

### CODE REVIEW

**Comment by blopker** [comment]
@C4ptainCrunch, any insight why this doesn't seem to be deploying?

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+          python -m pip install --upgrade pip
+          pip install -r requirements.txt
+      - name: Run unit tests
+        run: pytest
+      - name: Setup app
+        run: |
+          ./manage.py migrate
+          ./start.sh &
+      - name: Run integration tests
+        run: curl http://localhost:5000/
```
Comment (blopker) [inline .github/workflows/main.yml:24]: Adding CI with a simple integration test.

File: Procfile [hunk 2]
```diff
@@ -1,2 +1,2 @@
-web: uvicorn --limit-max-requests 10000 --limit-concurrency 1000 --host=0.0.0.0 --port=${PORT:-5000} realworld.asgi:application
+web: ./start.sh
```
Comment (blopker) [inline Procfile:1]: Moved to a script so we can test it in CI as well.

File: articles/views.py [hunk 3]
```diff
@@ -5,14 +5,79 @@
...
+
+        if favorited := self._get_query_context()["favorited"]:
+            qs = qs.filter(favorited_by__user__username=favorited)
+
+        return qs
+
+    def get_context_data(self, **kwargs):
+        ctx = super().get_context_data(**kwargs)
+        ctx.update(self._get_query_context())
+        return ctx
```
Comment (blopker) [inline articles/views.py:134]: This list view can filter based on a few different query params activated by the parent page's tabs.

File: home/views.py [hunk 4]
```diff
@@ -1,24 +1,35 @@
...
+    tag_slug = request.GET.get("tag")
+    tag = None
+    active_tab = request.GET.get("active_tab", "global")
+    article_params = ""
+    if tag_slug:
+        tag = get_object_or_404(Tag, slug=tag_slug)
+        active_tab = "tag"
+        article_params = f"tag={tag_slug}"
+    if active_tab == "feed":
+        article_params = f"followed_by={request.user.username}"
```
Comment (blopker) [inline home/views.py]: The tag selector is unique to the index page, but can control what is shown in the list frame. It does this by adding a query param to the index URL and reloading the page.

Comment (blopker) [reply home/views.py]: The index page also has 2 tabs, one for `My Feed` and one for `Global`, also activated by query params.

File: profiles/views.py [hunk 5]
```diff
@@ -8,12 +8,15 @@
...
-    context = {"article_user": user, "articles": articles}
+    active_tab = req.GET.get("active_tab", "my_articles")
+    article_params = f"author={user.username}"
+    if active_tab == "favorited":
+        article_params = f"favorited={user.username}"
+    context = {
+        "article_user": user,
+        "active_tab": active_tab,
+        "article_params": article_params,
+    }
```
Comment (blopker) [inline profiles/views.py:13]: Similarly set up to the index view, but with different list filter tabs.

File: templates/articles/list.html [hunk 6]
```diff
@@ -1,28 +1,76 @@
...
+        href="{% url 'article_list' %}?page={{ page_obj.next_page_number }}&{{query_string}}"
+        >next</a
+      >
+      <a
+        onClick="toTop()"
+        href="{% url 'article_list' %}?page={{ page_obj.paginator.num_pages }}&{{query_string}}"
+        >last &raquo;</a
+      >
+      {% endif %}
+    </span>
```
Comment (blopker) [inline templates/articles/list.html:83]: Added pagination controls. This is just lifted from the Django docs. Not super pretty, but works. `query_string` is used to preserve the current filtering while moving through the pages of results. Not sure if there's a more elegant way to persist this state.

File: templates/articles/list.html [hunk 7]
```diff
@@ -1,28 +1,76 @@
...
+    <script>
+      function toTop() {
+        window.scrollTo(0, 0);
+      }
```
Comment (blopker) [inline templates/articles/list.html:87]: Added a little JS so the user is move to the top of the page when a new page of results is requested. I could move this into an external JS file.

File: templates/articles/list.html [hunk 8]
```diff
@@ -1,28 +1,76 @@
-{% for article in articles %}
-<div class="article-preview">
-  <div class="article-meta">
-    <a href="{% url 'profile_view' profile=article.author.user.username %}"
-      ><img src="{{ article.author.img }}"
-    /></a>
-    <div class="info">
+<turbo-frame id="article-list">
```
Comment (blopker) [inline templates/articles/list.html:1]: Lazy turbo frame!

File: templates/articles/list.html [hunk 9]
```diff
@@ -1,28 +1,76 @@
...
+        <i class="ion-heart"></i> {{ article.favorited_by__count }}
+      </button>
     </div>
-    <button class="btn btn-outline-primary btn-sm pull-xs-right">
-      <i class="ion-heart"></i> {{ article.favorited_by__count }}
-    </button>
+    <a
+      href="{% url 'article_view' article.slug %}"
+      class="preview-link"
+      target="_top"
```
Comment (blopker) [inline templates/articles/list.html:41]: Using `target` here to make these links change top level navigation.

File: templates/index.html [hunk 10]
```diff
@@ -14,25 +14,46 @@ <h1 class="logo-font">conduit</h1>
           <ul class="nav nav-pills outline-active">
             {% if user.is_authenticated %}
             <li class="nav-item">
-              <a class="nav-link" href="">Your Feed</a>
+              <a
+                class="nav-link {% if active_tab == 'feed' %}active{% endif %}"
+                href="{% url 'index' %}?active_tab=feed"
```
Comment (blopker) [inline templates/index.html]: Tab switching is all done through query params and view code. No JS!

File: templates/index.html [hunk 11]
```diff
@@ -14,25 +14,46 @@ <h1 class="logo-font">conduit</h1>
...
+                >Global Feed</a
+              >
             </li>
+            {% if tag %}
+            <li class="nav-item">
+              <a class="nav-link {% if active_tab == 'tag' %}active{% endif %}"
+                >{{tag}}</a
+              >
+            </li>
+            {% endif %}
```
Comment (blopker) [inline templates/index.html]: The `tag` tab isn't actually clickable and just disappears once that filtering goes away.

File: templates/index.html [hunk 12]
```diff
@@ -14,25 +14,46 @@ <h1 class="logo-font">conduit</h1>
...
           </ul>
         </div>
-
-        {% include 'articles/list.html' %}
+        <turbo-frame
+          id="article-list"
+          src="{% url 'article_list'%}?{{article_params}}"
+        >
+        </turbo-frame>
```
Comment (blopker) [inline templates/index.html:14]: The article list lazy frame! The parent view translates the parent's query params into the correct params to send to the list view.

File: templates/profile/detail.html [hunk 13]
```diff
@@ -19,18 +19,29 @@ <h4>{{ article_user.username }}</h4>
...
+                href="{% url 'profile_view' profile=article_user.profile %}"
+                >My Articles</a
+              >
             </li>
             <li class="nav-item">
-              <a class="nav-link" href="">Favorited Articles</a>
+              <a
+                class="nav-link {% if active_tab == 'favorited' %}active{% endif %}"
+                href="{% url 'profile_view' profile=article_user.profile %}?active_tab=favorited"
+                >Favorited Articles</a
```
Comment (blopker) [inline templates/profile/detail.html]: Different tabs/query params from index page, but same concept.

File: articles/views.py [hunk 14]
```diff
@@ -5,14 +5,79 @@
...
+            ctx["favorited"] = favorited
+            ctx["query_string"].append(f"favorited={favorited}")
+
+        followed_by = self.request.GET.get("followed_by")
+        if followed_by:
+            ctx["followed_by"] = followed_by
+            ctx["query_string"].append(f"followed_by={followed_by}")
+
+        ctx["query_string"] = "&".join(ctx["query_string"])
+        return ctx
```
Comment (blopker) [inline articles/views.py:106]: Not super DRY, but easy enough to follow.

File: templates/base.html [hunk 15]
```diff
@@ -18,6 +18,7 @@
     <!-- Import the custom Bootstrap 4 theme from our hosted CDN -->
     <link rel="stylesheet" href="https://demo.productionready.io/main.css" />
 
+    <meta name="turbo-cache-control" content="no-preview" />
```
Comment (blopker) [inline templates/base.html]: Disabling previews since it causes flashing when revisiting article lists you've seen before.

**Comment by blopker** [comment]
OK, this is ready for review. I've refactored it a lot because I was unhappy with the previous implementation. I've learned doing a full page load with lazy frames is not a good user experience especially when the server is slow. The current code brings the tabs into the list frame and is now controlled by the list view instead of by the parent view. This makes the list view a little more complicated, but keeps all the code is one place which I like.

Sorry for the churn!

File: templates/index.html [hunk 16]
```diff
@@ -10,29 +10,21 @@ <h1 class="logo-font">conduit</h1>
...
-
           <div class="tag-list">
             {% for tag in tags %}
-            <a href="{{ tag.slug }}" class="tag-pill tag-default">{{ tag }}</a>
+            <a
+              href="{% url 'article_list' %}?tag={{ tag.slug }}"
+              class="tag-pill tag-default"
+              data-turbo-frame="article-list"
+              >{{ tag }}</a
+            >
```
Comment (blopker) [inline templates/index.html:27]: The tags use data-turbo-frame to target changes on the article-list frame.

**Comment by blopker** [comment]
Going to merge for now. Please leave comments if you have questions or improvements to make!

File: articles/views.py (hunk 3 above)
Comment (chrisgrande) [reply articles/views.py:134]: Could probably add some model managers to Article for the different use cases, especially if we need them in other views.

**Review by chrisgrande — COMMENTED** [review]
Looks good 👍🏻

File: articles/views.py (hunk 3 above)
Comment (blopker) [reply articles/views.py:134]: Ah, that's a good idea! We don't need them right now, but would clean up this view a bit. I'm taking over the "favorite" PR right now. I'm going to see if that will help there.

**Pull request merged**